### PR TITLE
Remove ZeroMQ dependency from master publisher

### PR DESCRIPTION
**Pull request by cachedout**
This pull request allows a Salt master to run with the TCP transport without depending on ZeroMQ for push/pull sockets inside the Publisher.

To accomplish this, we now have extended Tornado to implement a pure-pyhon UNIX socket-based IPC server/client. Asynchronous event handling and callback handling is controlled by Tornado's IOLoop.

There are still a few lingering issues to be solved, but I would like to see how ths runs across multiple platforms.

Refs #24612

### CODE REVIEW

**Comment by cachedout** [comment]
Go Go Jenkins!

**Comment by thatch45** [comment]
Wow, this looks very close!

**Comment by cachedout** [comment]
Yes, it should be. We're leaking filehandles somewhere but I haven't chased it down yet.

**Comment by jacksontj** [comment]
@cachedout This is "leaking" because there is a mix of sync and async calls ;) The primary ones are connect/write on the IOStreams. These both return futures (or callbacks-- which you seem to be using a lot of). Since nothing is waiting on the connect to happen before 'send' we are just queueing up a send after the connect. Then its a race condition-- which also means that we won't close the FD until the writes have all gone out.

I have a patch in my copy of your fork-- although github won't let me make a PR to your repo for some reason... so feel free to grab it ;)

Basically we need to be a little more cautious about blocking vs non-blocking calls. Also, I'd suggest standardizing on coroutine instead of callbacks-- since your code is primarily callbacks whereas the core is primarily coroutines :)

**Comment by cachedout** [comment]
@jacksontj Thanks for reviewing this.Yes, this is definietely messy and I appreciate the cleanup very much. I have been abroad for the past several weeks but I will try to get your fixes merged in to my branch tonight.

**Comment by jacksontj** [comment]
Awesome, let me know once you want another look over :)

Also, before merging we'll probably want to squash some of the commits (since there'll be a refactor). You'll be *much* happier with that if there is an issue that you have to bisect through this code :)

**Comment by cachedout** [comment]
@jacksontj (I'm back in the office now). I pushed up a few small changes today with a few of your in-line recommendations. They were helpful, thank you! Something is still awry with one of the transport tests so I'm chasing that down and will then start migrating over to coroutines.

**Comment by cachedout** [comment]
Go Go Jenkins!

**Comment by cachedout** [comment]
Go Go Jenkins!

**Comment by cachedout** [comment]
@jacksontj I think we're just about ready to get this merged. Let me know when you have a moment to go over it before we get it in. We can meet up in IRC if you want as well. Thanks.

**Comment by jacksontj** [comment]
I should be hanging out on IRC for the rest of the day :)

File: salt/transport/ipc.py [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+            self.io_loop.spawn_callback(self.stream_handler, body)
+
+    @tornado.gen.coroutine
+    def handle_connection(self, connection, address):
+        log.trace('IPCServer: Handling connection to address: {0}'.format(address))
+        try:
+            stream = IOStream(connection,
+                              io_loop=self.io_loop,
+                              )
+            yield self.handle_stream(stream)
```
Comment (jacksontj) [inline salt/transport/ipc.py]: IIRC this means you'll do exactly one stream at a time. I think you want to spawn the callback, so you can have multiple in flight at once. (which would mean this method doesn't need to be a coroutine)

File: salt/transport/ipc.py [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+        IPC clients cannot bind to ports, but must connect to
+        existing IPC servers. Clients can then send messages
+        to the server.
+
+        '''
+        self.io_loop = io_loop or tornado.ioloop.IOLoop.current()
+        self.socket_path = socket_path
+
+    # TODO: single connect for multiple callers (similar to auth)
+    @tornado.gen.coroutine
```
Comment (jacksontj) [inline salt/transport/ipc.py]: Might want to so a singleton-esque connect-- that way if multiple people call methods that would do connect we only *actually* do one. If you want an example the auth system does this.

**Comment by cachedout** [comment]
@jacksontj I made the changes you mentioned the other day. I rebased into a single commit and would like to get this in today so we can have QA start looking it over. Let me know if you have any strong objections. There will still be plenty of time to refine this before the release of course. :]

File: tests/unit/transport/ipc.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+#        print('Received payload: {0}'.format(payload))
+#        raise tornado.gen.Return((payload, {'fun': 'send_ipc'}))
+#
+#    def test_simple_push(self):
+#        self.push_channel.send('Hello world')
+
+    @tornado.testing.gen_test
+    def test_async(self):
+        pull_channel = salt.transport.ipc.IPCServer({}, io_loop=self.io_loop)
+        pull_channel.start('/tmp/t.ipc')
```
Comment (jacksontj) [inline tests/unit/transport/ipc.py]: We should make sure the files generated are contained in the regular test tmp dirs.

File: tests/unit/transport/ipc.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+        push_channel = salt.transport.ipc.IPCMessageClient({}, socket_path='/tmp/t.ipc', io_loop=self.io_loop)  # pylint: disable=E1124
+        push_channel.connect()
+        push_channel.send('Hello world')
```
Comment (jacksontj) [inline tests/unit/transport/ipc.py]: Definitely need some sort of asserts in here :) I'd also like to see some concurrency testing, since this is such a critical part of the core we should have good coverage (which would also include negative cases).

File: salt/transport/tcp.py [hunk 5]
```diff
@@ -647,9 +591,10 @@ def publish_payload(self, package):
 
 
 class TCPPubServerChannel(salt.transport.server.PubServerChannel):
-    def __init__(self, opts):
+    def __init__(self, opts, io_loop=None):
         self.opts = opts
         self.serial = salt.payload.Serial(self.opts)  # TODO: in init?
+        self.io_loop = io_loop or tornado.ioloop.IOLoop.current()
 
     def _publish_daemon(self):
```
Comment (jacksontj) [inline salt/transport/tcp.py]: Since this is now on the IOLoop we don't even need a separate process for this. If we want to maintain one we should keep our own ioloop instead of start() ing one from the base process to avoid weirdness

File: salt/transport/ipc.py [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+    def handle_stream(self, stream):
+        '''
+        Override this to handle the streams as they arrive
+
+        :param IOStream stream: An IOStream for processing
+
+        See http://tornado.readthedocs.org/en/latest/iostream.html#tornado.iostream.IOStream
+        for additional details.
+        '''
+        while True:
```
Comment (jacksontj) [inline salt/transport/ipc.py]: We probably want some try/except wrapped in a close. If the client sends us garbage we should log an error and close their stream. Right now we'll stacktrace :)

File: salt/transport/ipc.py [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+    '''
+    def __init__(self,
+                 opts,
+                 socket_path=None,
+                 io_loop=None):
+        '''
+        Create an IPCMessageClient
+        '''
+        super(IPCMessageClient, self).__init__(io_loop=io_loop, socket_path=socket_path)
+        self.io_loop = io_loop or tornado.ioloop.IOLoop.current()
```
Comment (jacksontj) [inline salt/transport/ipc.py]: Parent constructor does both of these?

File: salt/transport/ipc.py [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+        if hasattr(self, 'stream'):
+            self.stream.close()
+
+    def __del__(self):
+        '''
+        Routines to handle any cleanup before the instance shuts down.
+        Sockets and filehandles should be closed explicitely, to prevent
+        leaks.
+        '''
+        if hasattr(self, 'stream'):
```
Comment (jacksontj) [inline salt/transport/ipc.py]: Should just call self.close()

**Comment by jacksontj** [comment]
There are also a few test failures that look like they are probably related

**Comment by jacksontj** [comment]
Is the plan to get this in for 2015.8? If so we really need to get this in soon since the RC was just cut ;) Do you need help finishing this up?

**Comment by cachedout** [comment]
I just pushed a (hopefully) better set of test cases but the TCP test is still failing and I haven't had a chance to figure out why. I'd certainly welcome help, since this is a targeted (but not required) feature for 2015.8.

**Comment by jacksontj** [comment]
@cachedout I have a PR open to your branch which fixes these all up. We can get this in for 2015.8, but if we do we should probably do all the IPC (pub/sub and push/pull), which shouldn't be much work, but we don't really want to split the IPC up ;) When is 2015.8 slated for release (I presume next month)?

**Comment by cachedout** [comment]
Yes, slated for next month. My thinking was that we would do this and then pub/sub in a release after that but maybe we miscommunicated? (Totally possible I misunderstood.) We can certainly try and get it in. @thatch45 what do you think?

**Comment by jacksontj** [comment]
IMO if we are going to get one in, we might as well do all 3. pub/sub is different, but push/pull is basically the same as the TCP message client/server (they support send/recv). I might be able to make some time later this week to add that push/pull functionality. If we consolidate the tcp implementation as well it already supports all 3 with the same client.

That being said we are getting aweful close to the release, when is the next feature release after 2015.8?

**Comment by cachedout** [comment]
We can get additional features into the first point release, targeted for two weeks after Beryllium.

**Comment by cachedout** [comment]
@jacksontj I'm ready to merge this if you are. :]

**Comment by jacksontj** [comment]
If the test are passing, let's do it

**Comment by cachedout** [comment]
Done!

**Comment by s0undt3ch** [comment]
:clap:
